copylot/hardware/lasers/vortran/test/test_vortran.py

```diff
@@ -1,49 +1,160 @@
+    that are expected by VortranLaser._identify_laser
+    '''
+    mock_serial = mocker.patch("serial.Serial", autospec=True)
+    mock_serial_instance = mock_serial.return_value
+
+    # the value returned by calls to readline() during instantiation (i.e., in `_identify_laser`)
+    mock_serial_instance.readline.return_value = (
+        f"?LI={serial_number}, 600nm, 400mW, 500mW, laser-shape\r\n".encode()
+    )
+    # the `is_open` attribute is not found by autospec, so we manually define it
+    mock_serial_instance.is_open = True
+
+    return mock_serial
+
+
+@pytest.fixture
+def mock_ports(mocker):
+    '''
+    Mock the list_ports.comports function and return the list of ports
+    '''
+    mock_list_ports = mocker.patch("serial.tools.list_ports.comports", autospec=True)
+    mock_ports = [mock.MagicMock(device=port) for port in ['some-port', 'another-port']]
+    mock_list_ports.return_value = mock_ports
+    return mock_ports
+
+
+@pytest.fixture
+def create_laser(mock_serial):
+    '''
+    Factory fixture to create a VortranLaser instance
+    with mocked return values from one or more post-instantiation calls to `readline`
+    (these calls are triggered by calls to the getters/setters)
+    '''
+
+    def _create_laser(readline_return_values):
+        mock_serial.return_value.readline.side_effect = [
+            mock_serial.return_value.readline.return_value,
+            *[value.encode() for value in readline_return_values],
+        ]
+        laser = VortranLaser(port="some-port")
+        return laser
+
+    return _create_laser
+
 
+def test_connect_with_serial_number(mock_serial, mock_ports, serial_number):
+    '''
+    if the user provides a serial number, available COM ports are tried
+    until a laser is found that matches the provided serial number
 
-def demo_toggle_emission():
-    """
-    Test toggling the emission
-    """
-    laser_list = vortran.VortranLaser.get_lasers()
-    laser = vortran.VortranLaser(port=laser_list[0][0])
-    laser.set_emission_delay = 0
-    laser.laser_power = 2.0
-    laser.toggle_emission = 1
-    time.sleep(3)
-    laser.toggle_emission = 0
+    note: since we've mocked the serial.Serial().readline function,
+    we know that the laser at the first port will match the serial_number
+    '''
+    laser = VortranLaser(serial_number=serial_number)
+
+    mock_serial.assert_called_once_with(
+        port=mock_ports[0],
+        baudrate=laser.baudrate,
+        bytesize=serial.EIGHTBITS,
+        parity=serial.PARITY_NONE,
+        stopbits=serial.STOPBITS_ONE,
+        timeout=laser.timeout,
+    )
+    assert laser.is_connected
+    assert laser.serial_number == serial_number
+    assert laser.address == mock_serial.return_value
+
+
+@pytest.mark.usefixtures('mock_serial')
+def test_connect_with_wrong_serial_number():
+    '''
+    if a serial number is provided, it must be correct or else the laser is not connected
+    TODO: should this raise an exception?
```

Comment:
I was in debate with this too. Considering there are mainly three ways of connecting: 1) Ideally with the serial number since that does not change 2) COM Port number. There is a risk that when the computer restarts or cables are re-plugged in that the port numbers are reassigned 3) Serial + COM port, which is somewhat tied to 2.

Reply:
actually since this is the pytest this should raise a warning.